src/sinks/tcp.rs

```diff
@@ -106,30 +247,59 @@ impl TcpSink {
                             message = "disconnected.",
                             addr = &field::display(&self.addr)
                         );
-                        self.state = TcpSinkState::Disconnected;
+                        TcpSinkState::Disconnected
                     }
                 },
                 TcpSinkState::Connecting(ref mut connect_future) => match connect_future.poll() {
                     Ok(Async::Ready(socket)) => {
                         let addr = socket.peer_addr().unwrap_or(self.addr);
                         debug!(message = "connected", addr = &field::display(&addr));
-                        self.state =
-                            TcpSinkState::Connected(FramedWrite::new(socket, BytesCodec::new()));
                         self.backoff = Self::fresh_backoff();
+                        match self.tls {
+                            Some(ref tls) => match tls.make_connector() {
+                                Ok(connector) => TcpSinkState::TlsConnecting(
+                                    connector.connect(&self.hostname, socket),
+                                ),
+                                Err(err) => {
+                                    error!("{}", err);
+                                    TcpSinkState::Backoff(self.next_delay())
```

Comment:
should we be retrying on failed tls handshakes?

Reply:
What is the alternative? The TCP client doesn't have a "failed" state, and TLS handshakes can fail temporarily (memory pressure, interrupted connection, etc).